### PR TITLE
Minimal petsc4py code for ksp

### PR DESCRIPTION
**Pull request by enricofacca**
Minimal code using petsc4py and ksp solver for linear systems.

Test add to test_wasserstein

### CODE REVIEW

File: src/darsia/utils/linalg.py [hunk 1]
```diff
@@ -19,3 +19,255 @@ def __init__(self, A: sps.csc_matrix) -> None:
...
+        # This step is needed to avoid a petsc error with LU factorization
+        # that explicitly needs a zeros along the diagonal
+        # TODO: use zero_diagonal = PETSc.Mat().createConstantDiagonal(self.A.size,0.0)
+        zero_diagonal = PETSc.Mat().createAIJ(
+            size=A.shape, 
+            csr=(np.arange(A.shape[0]+1,dtype="int32"),
+                np.arange(A.shape[0],dtype="int32"),
+                np.zeros(A.shape[0]))
+        )
+        self.A_petsc += zero_diagonal
```
[the remark this answers is not in the document]
Comment (jwboth) [reply src/darsia/utils/linalg.py:57]: Less prioritized for now.

File: src/darsia/utils/linalg.py [hunk 2]
```diff
@@ -19,3 +19,255 @@ def __init__(self, A: sps.csc_matrix) -> None:
...
+
+        self.appctx = appctx
+        
+        
+
+    def setup(self, petsc_options: dict) -> None:
+        petsc_options = flatten_parameters(petsc_options)
+
+
+        self.prefix = "petsc_solver_"
```
Comment (jwboth) [inline src/darsia/utils/linalg.py:90]: Does this has implications for how to define the options of the solver from the outside?

File: requirements.txt [hunk 3]
```diff
@@ -13,3 +13,4 @@ pandas
 plotly
 largestinteriorrectangle
 pydicom
+petsc4py
```
Comment (jwboth) [inline requirements.txt:16]: Based on the github actions, github does not manage to install DarSIA including petsc4py. Can you have a look at the installation report? Anything needed in addition?

**Review by jwboth — CHANGES_REQUESTED** [review]
Thanks a lot, Enrico! This looks really good. Most comments are curiosity questions, me wanting to better understand what is going on. 

Most critical is the failing installation of DarSIA including petsc4py. The style (black etc.) will sort out. I can take care of that once the installation is fixed.

File: tests/unit/test_wasserstein.py [hunk 4]
```diff
@@ -118,7 +118,23 @@
...
+    "linear_solver_options": {
+        "tol": 1e-8,
+        "approach": "direct",
+    },
+}
+ksp_krylov_options = {
+    "linear_solver": "ksp",
+    "linear_solver_options": {
+        "tol": 1e-8,
+        "approach": "krylov",
```
[the remark this answers is not in the document]
Comment (enricofacca) [reply tests/unit/test_wasserstein.py:132]: Yes

File: src/darsia/measure/wasserstein.py [hunk 5]
```diff
@@ -1069,6 +1115,12 @@ def linear_solve(
 
                 elif self.linear_solver_type == "cg":
                     self.setup_cg_solver(self.fully_reduced_matrix)
+                
+                elif self.linear_solver_type == "ksp":
+                    # Free memory if solver needs to be re-setup
+                    if not (reuse_solver) and hasattr(self, "linear_solver"):
+                        self.linear_solver.kill()
+                    self.setup_ksp_solver(self.fully_reduced_matrix)
```
[the remark this answers is not in the document]
Comment (enricofacca) [reply src/darsia/measure/wasserstein.py:1123]: Yes. Need to be fixed.

**Comment by enricofacca** [comment]
Some comments on why we are not able to install petsc4py.
Reading [this ](https://pypi.org/project/petsc4py/) we first need to install numpy (than has recently being updated to version 2.0) and mpi4py. 
Than we should install petsc and petsc4py via pip.
I was able to make it work but mumps and hypre are missing. 
It should be possible to include this component using environment variable (as described [here](https://gitlab.com/petsc/petsc/-/blob/main/setup.py?ref_type=heads)) but it does not work.

I was able to make everything work using conda + create env + installing petsc with conda install (pip install inside the env does not work). Hypre and mumps are, by default, included in the conda installation of petsc.

**Comment by enricofacca** [comment]
It seems that the issue with the inclusion of petsc4py is not so easy to solve.
Following the instruction [here](https://pypi.org/project/petsc4py/) I was was to include petsc4py.
However hypre and mumps are missing. [Here](https://gitlab.com/petsc/petsc/-/blob/main/setup.py?ref_type=heads) it says that it possible to include them using environment variables, but I was able to.

Nevertheless using conda + create env + installing via conda (and not via pip within the env) I was able to install petsc, hypre and mumps , since they are included by default.

**Comment by jwboth** [comment]
I removed petsc4py and cython from the requirements. Instead, now petsc4py (at least on github actions) is build in the build workflow, see #374 . The installation for Windows will not fully work from command line I fear, and not just using pip. I am trying therefore instead to set up a separate routine for that. Can you check, whether you can run the installation of petsc incl. hypre and mumps as done in the ci script?

**Review by enricofacca — COMMENTED** [review]
I answered your points. But I just wanted to test the CI. I am not sure I did the review as it supposed to be.

**Comment by enricofacca** [comment]
Testing works but it took **"30 minutes"** to complete with the inclusion of hypre and mumps. 
This may be a problem if there are limitations on free testing time on Github.

We can just use native petsc solvers (replacing "hypre" with "icc" and mumps with the standard LU factorization), include petsc.  So that the building test is passed and it takes way less time.

However, the installation of mpi4py requires MPI being installed. This may cause problems to some users..

**Review by jwboth — APPROVED** [review]
Looks good for now! Thanks a lot!

**Pull request merged**